File: backend/app/services/training_engine.py
from datetime import date, datetime, timedelta
from typing import Optional
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, desc
from app.models.user import User
from app.models.activity import Activity
from app.models.competition import Competition
from app.models.training_session import TrainingSession, SessionSource, SessionStatus
from app.core.claude_client import claude_client
from app.prompts.training_recommendation import (
    TRAINING_RECOMMENDATION_SYSTEM,
    TRAINING_RECOMMENDATION_PROMPT,
    PLAN_CONVERSION_SYSTEM,
    PLAN_CONVERSION_PROMPT,
)
from app.services.strava_service import format_pace


async def generate_recommendations(
    user: User,
    db: AsyncSession,
    start_date: date,
    end_date: date,
    consider_fixed_plan: bool = True,
    include_cross_training: bool = True,
) -> dict:
    """
    Generate training recommendations for a user for the given date range.
    """
    # Get recent activities (last 30 days for better context)
    thirty_days_ago = datetime.utcnow() - timedelta(days=30)
    result = await db.execute(
        select(Activity)
        .where(Activity.user_id == user.id)
        .where(Activity.start_date >= thirty_days_ago)
        .order_by(desc(Activity.start_date))
    )
    recent_activities = result.scalars().all()

    # Calculate weekly stats (last 7 days)
    seven_days_ago = datetime.utcnow() - timedelta(days=7)
    week_activities = [a for a in recent_activities if a.start_date >= seven_days_ago]

    weekly_distance = sum(a.distance or 0 for a in week_activities) / 1000
    weekly_duration = sum(a.duration or 0 for a in week_activities) / 3600
    weekly_avg_hr = sum(a.avg_heart_rate or 0 for a in week_activities) / max(len(week_activities), 1)

    # Count sessions by sport type
    run_types = {"run", "running", "trail run", "treadmill"}
    ride_types = {"ride", "cycling", "virtualride", "virtual ride"}
    weekly_runs = sum(1 for a in week_activities if (a.activity_type or "").lower() in run_types)
    weekly_rides = sum(1 for a in week_activities if (a.activity_type or "").lower() in ride_types)
    weekly_other = len(week_activities) - weekly_runs - weekly_rides

    # Get all competitions within the planning period (and slightly beyond for taper planning)
    result = await db.execute(
        select(Competition)
        .where(Competition.user_id == user.id)
        .where(Competition.race_date >= start_date)
        .where(Competition.race_date <= end_date + timedelta(days=14))  # Include races slightly after end
        .order_by(Competition.race_date)
    )
    upcoming_competitions = result.scalars().all()

    # Get fixed plan sessions for the date range
    fixed_plan_text = "No fixed training plan."
    if consider_fixed_plan:
        result = await db.execute(
            select(TrainingSession)
            .where(TrainingSession.user_id == user.id)
            .where(TrainingSession.session_date >= start_date)
            .where(TrainingSession.session_date <= end_date)
            .where(TrainingSession.planned_workout.isnot(None))
            .order_by(TrainingSession.session_date)
        )
        fixed_sessions = result.scalars().all()
        if fixed_sessions:
            fixed_plan_text = "\n".join([
                f"- {s.session_date}: {s.planned_workout.get('description', 'No description')}"
                for s in fixed_sessions
            ])

    # Format activities for prompt (show up to 15 activities for better context)
    activities_text = "No recent activities." if not recent_activities else "\n".join([
        f"- {a.start_date.strftime('%Y-%m-%d')}: [{a.activity_type or 'Run'}] {a.name} - "
        f"{(a.distance or 0)/1000:.1f}km, "
        f"{format_pace(a.avg_pace)}/km, "
        f"HR: {a.avg_heart_rate or 'N/A'}bpm"
        for a in recent_activities[:15]
    ])

    # Format competitions for prompt with days until race
    if not upcoming_competitions:
        competitions_text = "No upcoming competitions."
    else:
        comp_lines = []
        for c in upcoming_competitions:
            days_until = (c.race_date - start_date).days
            # Handle enum values safely (may be string or enum)
            race_type = c.race_type.value if hasattr(c.race_type, 'value') else str(c.race_type)
            priority = c.priority.value if hasattr(c.priority, 'value') else str(c.priority)
            comp_lines.append(
                f"- {c.race_date}: {c.name} ({race_type}) - "
                f"Priority: {priority}, "
                f"Goal: {format_goal_time(c.goal_time)}, "
                f"Days until: {days_until}"
            )
        competitions_text = "\n".join(comp_lines)

    # Format zones
    preferences = user.preferences or {}
    hr_zones = preferences.get("hr_zones", {})
    pace_zones = preferences.get("pace_zones", {})

    hr_zones_text = "\n".join([
        f"- {zone}: {data.get('min', 0)}-{data.get('max', 0)} bpm ({data.get('name', zone)})"
        for zone, data in hr_zones.items()
    ]) or "Default zones"

    pace_zones_text = "\n".join([
        f"- {zone}: {format_pace(data.get('min', 0))}-{format_pace(data.get('max', 0))}/km ({data.get('name', zone)})"
        for zone, data in pace_zones.items()
    ]) or "Default zones"

    # Format cycling power zones
    ftp = preferences.get("ftp")
    cycling_power_zones = preferences.get("cycling_power_zones", {})

    ftp_text = f"{ftp}W" if ftp else "Not set"

    cycling_power_zones_text = "\n".join([
        f"- {zone}: {data.get('min', 0)}-{data.get('max', 0)}W ({data.get('name', zone)})"
        for zone, data in cycling_power_zones.items()
    ]) or "Not configured (no FTP set)"

    # Calculate planning duration
    planning_days = (end_date - start_date).days
    planning_weeks = planning_days // 7

    # Get athlete profile summary
    athlete_profile = user.profile_summary or "No profile summary available."

    # Get threshold pace for context
    threshold_pace = preferences.get("threshold_pace")
    threshold_pace_text = format_pace(threshold_pace) if threshold_pace else "Not set"

    # Build prompt
    prompt = TRAINING_RECOMMENDATION_PROMPT.format(
        athlete_name=user.name,
        max_hr=preferences.get("max_hr", 190),
        resting_hr=preferences.get("resting_hr", 50),
        threshold_pace=threshold_pace_text,
        ftp=ftp_text,
        athlete_profile=athlete_profile,
        hr_zones=hr_zones_text,
        pace_zones=pace_zones_text,
        cycling_power_zones=cycling_power_zones_text,
        recent_activities=activities_text,
        weekly_distance=f"{weekly_distance:.1f}",
        weekly_duration=f"{weekly_duration:.1f}",
        weekly_avg_hr=f"{weekly_avg_hr:.0f}",
        weekly_runs=weekly_runs,
        weekly_rides=weekly_rides,
        weekly_other=weekly_other,
        upcoming_competitions=competitions_text,
        fixed_plan=fixed_plan_text,
        start_date=start_date.isoformat(),
        end_date=end_date.isoformat(),
        planning_weeks=planning_weeks,
    )

    if not include_cross_training:
        prompt += (
            "\n\nIMPORTANT: The athlete has requested a RUNNING-ONLY plan. "
            "Do NOT include cross-training sessions (cycling, swimming, strength, etc.). "
            "All sessions should have sport='running'. Use rest days instead of cross-training days."
        )

    # Call Claude
    result = await claude_client.generate_training_recommendations(
        TRAINING_RECOMMENDATION_SYSTEM,
        prompt
    )

    return result


async def convert_session(
    user: User,
    workout: dict,
    target_type: str,  # "hr_based" or "pace_based"
) -> dict:
    """
    Convert a training session between pace-based and HR-based formats.
    """
    preferences = user.preferences or {}
    hr_zones = preferences.get("hr_zones", {})
    pace_zones = preferences.get("pace_zones", {})

    source_type = "pace-based" if target_type == "hr_based" else "HR-based"

    hr_zones_text = "\n".join([
        f"- {zone}: {data.get('min', 0)}-{data.get('max', 0)} bpm ({data.get('name', zone)})"
        for zone, data in hr_zones.items()
    ])

    pace_zones_text = "\n".join([
        f"- {zone}: {format_pace(data.get('min', 0))}-{format_pace(data.get('max', 0))}/km ({data.get('name', zone)})"
        for zone, data in pace_zones.items()
    ])

    import json
    session_details = json.dumps(workout, indent=2)

    prompt = PLAN_CONVERSION_PROMPT.format(
        source_type=source_type,
        target_type=target_type.replace("_", " "),
        hr_zones=hr_zones_text,
        pace_zones=pace_zones_text,
        session_details=session_details,
        workout_type=workout.get("type", "easy"),
    )

    result = await claude_client.convert_session(
        PLAN_CONVERSION_SYSTEM,
        prompt
    )

    return result


async def save_recommendations(
    user: User,
    db: AsyncSession,
    recommendations: dict
) -> list[TrainingSession]:
    """
    Save generated recommendations to the database.
    Only saves to future days (including today). Always replaces AI recommendation column.
    """
    sessions = recommendations.get("sessions", [])
    saved_sessions = []
    today = date.today()

    for session_data in sessions:
        session_date_str = session_data.get("date")
        if not session_date_str:
            continue

        session_date = date.fromisoformat(session_date_str)

        # Only process future days (including today)
        if session_date < today:
            continue

        # Check if session already exists for this date
        result = await db.execute(
            select(TrainingSession)
            .where(TrainingSession.user_id == user.id)
            .where(TrainingSession.session_date == session_date)
        )
        existing = result.scalar_one_or_none()

        if existing:
            # Always replace AI recommendation on existing session
            existing.recommendation_workout = session_data
        else:
            # Create new session
            session = TrainingSession(
                user_id=user.id,
                session_date=session_date,
                source=SessionSource.APP_RECOMMENDATION,
                recommendation_workout=session_data,
            )
            db.add(session)
            saved_sessions.append(session)

    await db.commit()
    return saved_sessions


def format_goal_time(seconds: Optional[int]) -> str:
    """Format goal time from seconds to HH:MM:SS or MM:SS"""
    if not seconds:
        return "N/A"
    hours = seconds // 3600
    minutes = (seconds % 3600) // 60
    secs = seconds % 60

    if hours > 0:
        return f"{hours}:{minutes:02d}:{secs:02d}"
    return f"{minutes}:{secs:02d}"

File: src/server/services/training-engine.ts
import "server-only";
import { and, asc, desc, eq, gte, isNotNull, lte } from "drizzle-orm";
import { db } from "@/server/db";
import {
  activities,
  competitions,
  trainingSessions,
  type Activity,
  type Competition,
  type TrainingSession,
  type User,
  type UserPreferences,
} from "@/server/db/schema";
import { calculateTrimp } from "@/server/services/training-load";
import { resolveEffectiveWorkout } from "@/server/services/workout-normalize";
import {
  convertSession as claudeConvert,
  generateTrainingRecommendations,
  type GenerationProgressCallback,
} from "@/server/services/claude";
import { formatGoalTime, formatPace } from "@/server/services/pace";
import {
  buildPlanConversionPrompt,
  buildTrainingRecommendationPrompt,
  PLAN_CONVERSION_SYSTEM,
  TRAINING_RECOMMENDATION_SYSTEM,
} from "@/server/prompts/training-recommendation";

// ---------------------------------------------------------------------------
// Helpers for prompt formatting
// ---------------------------------------------------------------------------

const RUN_TYPES = new Set(["run", "running", "trail run", "treadmill"]);
const RIDE_TYPES = new Set(["ride", "cycling", "virtualride", "virtual ride"]);

interface ZoneEntry {
  min?: number;
  max?: number;
  name?: string;
}

function formatHrZones(hrZones: Record<string, ZoneEntry> | undefined): string {
  if (!hrZones || Object.keys(hrZones).length === 0) return "Default zones";
  return Object.entries(hrZones)
    .map(([zone, data]) =>
      `- ${zone}: ${data.min ?? 0}-${data.max ?? 0} bpm (${data.name ?? zone})`,
    )
    .join("\n");
}

function formatPaceZones(paceZones: Record<string, ZoneEntry> | undefined): string {
  if (!paceZones || Object.keys(paceZones).length === 0) return "Default zones";
  return Object.entries(paceZones)
    .map(([zone, data]) =>
      `- ${zone}: ${formatPace(data.min ?? 0)}-${formatPace(data.max ?? 0)}/km (${data.name ?? zone})`,
    )
    .join("\n");
}

function formatPowerZones(powerZones: Record<string, ZoneEntry> | undefined): string {
  if (!powerZones || Object.keys(powerZones).length === 0) return "Not configured (no FTP set)";
  return Object.entries(powerZones)
    .map(([zone, data]) =>
      `- ${zone}: ${data.min ?? 0}-${data.max ?? 0}W (${data.name ?? zone})`,
    )
    .join("\n");
}

function isoOf(d: Date): string {
  return d.toISOString().slice(0, 10);
}

function dateFromIso(iso: string): Date {
  return new Date(`${iso}T00:00:00Z`);
}

// ---------------------------------------------------------------------------
// generate_recommendations
// ---------------------------------------------------------------------------

export interface GenerateInput {
  user: User;
  startDate: string; // YYYY-MM-DD
  endDate: string;
  considerFixedPlan?: boolean;
  allowedSports?: string[] | null;
  sportAvailability?: Record<string, { start_date?: string }> | null;
}

export async function generateRecommendations(
  input: GenerateInput,
  onProgress?: GenerationProgressCallback,
): Promise<Record<string, unknown>> {
  const { user, startDate, endDate, considerFixedPlan = true, allowedSports = null, sportAvailability = null } = input;
  const prefs = (user.preferences ?? {}) as UserPreferences;

  // Recent activities (30 days)
  const cutoff30 = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000);
  const recent = await db
    .select()
    .from(activities)
    .where(and(eq(activities.userId, user.id), gte(activities.startDate, cutoff30)))
    .orderBy(desc(activities.startDate));

  // Weekly stats (7 days)
  const cutoff7 = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);
  const week = recent.filter((a) => a.startDate >= cutoff7);
  const weeklyDistance = week.reduce((s, a) => s + (a.distance ?? 0), 0) / 1000;
  const weeklyDuration = week.reduce((s, a) => s + (a.duration ?? 0), 0) / 3600;
  const weeklyAvgHr = week.length
    ? week.reduce((s, a) => s + (a.avgHeartRate ?? 0), 0) / week.length
    : 0;
  const weeklyRuns = week.filter((a) => RUN_TYPES.has((a.activityType ?? "").toLowerCase())).length;
  const weeklyRides = week.filter((a) => RIDE_TYPES.has((a.activityType ?? "").toLowerCase())).length;
  const weeklyOther = week.length - weeklyRuns - weeklyRides;

  // Upcoming competitions (within range, plus 14 days)
  const compsEnd = isoOf(new Date(dateFromIso(endDate).getTime() + 14 * 24 * 60 * 60 * 1000));
  const upcoming = await db
    .select()
    .from(competitions)
    .where(
      and(
        eq(competitions.userId, user.id),
        gte(competitions.raceDate, startDate),
        lte(competitions.raceDate, compsEnd),
      ),
    )
    .orderBy(asc(competitions.raceDate));

  // Fixed plan
  let fixedPlanText = "No fixed training plan.";
  if (considerFixedPlan) {
    const fixed = await db
      .select()
      .from(trainingSessions)
      .where(
        and(
          eq(trainingSessions.userId, user.id),
          gte(trainingSessions.sessionDate, startDate),
          lte(trainingSessions.sessionDate, endDate),
          isNotNull(trainingSessions.plannedWorkout),
        ),
      )
      .orderBy(asc(trainingSessions.sessionDate));
    if (fixed.length > 0) {
      fixedPlanText = fixed
        .map((s) => {
          const w = (s.plannedWorkout ?? {}) as Record<string, unknown>;
          const desc = typeof w.description === "string" ? w.description : "No description";
          return `- ${s.sessionDate}: ${desc}`;
        })
        .join("\n");
    }
  }

  // Format recent activities (top 15)
  const activitiesText =
    recent.length === 0
      ? "No recent activities."
      : recent
          .slice(0, 15)
          .map(
            (a: Activity) =>
              `- ${a.startDate.toISOString().slice(0, 10)}: [${a.activityType ?? "Run"}] ${a.name} - ` +
              `${((a.distance ?? 0) / 1000).toFixed(1)}km, ` +
              `${formatPace(a.avgPace)}/km, ` +
              `HR: ${a.avgHeartRate ?? "N/A"}bpm`,
          )
          .join("\n");

  // Competitions text
  const competitionsText =
    upcoming.length === 0
      ? "No upcoming competitions."
      : upcoming
          .map((c: Competition) => {
            const daysUntil = Math.round(
              (dateFromIso(c.raceDate as unknown as string).getTime() - dateFromIso(startDate).getTime()) /
                (24 * 60 * 60 * 1000),
            );
            return (
              `- ${c.raceDate}: ${c.name} (${c.raceType}) - ` +
              `Priority: ${c.priority}, ` +
              `Goal: ${formatGoalTime(c.goalTime)}, ` +
              `Days until: ${daysUntil}`
            );
          })
          .join("\n");

  // Zones
  const hrZonesText = formatHrZones(prefs.hr_zones);
  const paceZonesText = formatPaceZones(prefs.pace_zones);
  const powerZonesText = formatPowerZones(prefs.cycling_power_zones);
  const ftpText = prefs.ftp ? `${prefs.ftp}W` : "Not set";

  // Weekly TRIMP for last 4 weeks
  const restingHr = prefs.resting_hr ?? 50;
  const maxHr = prefs.max_hr ?? 190;
  const trimpLines: string[] = [];
  for (let off = 0; off < 4; off++) {
    const wEnd = new Date(Date.now() - off * 7 * 24 * 60 * 60 * 1000);
    const wStart = new Date(wEnd.getTime() - 7 * 24 * 60 * 60 * 1000);
    const wActs = recent.filter((a) => a.startDate >= wStart && a.startDate < wEnd);
    const wTrimp = wActs.reduce((s, a) => {
      if (!a.avgHeartRate || !a.duration) return s;
      return s + calculateTrimp(a.duration, a.avgHeartRate, restingHr, maxHr);
    }, 0);
    trimpLines.push(`- Week -${off + 1}: ${wTrimp.toFixed(0)} TRIMP (${wActs.length} sessions)`);
  }

  // Recent RPE feedback
  const rpeStart = isoOf(new Date(dateFromIso(startDate).getTime() - 14 * 24 * 60 * 60 * 1000));
  const rpeRows = await db
    .select()
    .from(trainingSessions)
    .where(
      and(
        eq(trainingSessions.userId, user.id),
        isNotNull(trainingSessions.rpeActual),
        gte(trainingSessions.sessionDate, rpeStart),
      ),
    )
    .orderBy(desc(trainingSessions.sessionDate))
    .limit(10);
  const rpeText =
    rpeRows.length === 0
      ? "No RPE feedback yet."
      : rpeRows
          .map((rs: TrainingSession) => {
            // Expanded: a stored recommendation carries `rpe`, not
            // `rpe_target`, so reading it raw told the coach "no target" for
            // every session it had itself prescribed.
            const workout = (resolveEffectiveWorkout(rs) ?? {}) as Record<string, unknown>;
            const target = workout.rpe_target;
            const targetStr = typeof target === "number" ? `target=${target}` : "no target";
            return `- ${rs.sessionDate}: actual=${rs.rpeActual} (${targetStr})`;
          })
          .join("\n");

  // Planning weeks
  const planningDays = Math.round(
    (dateFromIso(endDate).getTime() - dateFromIso(startDate).getTime()) / (24 * 60 * 60 * 1000),
  );
  const planningWeeks = Math.floor(planningDays / 7);

  const thresholdPaceText = prefs.threshold_pace ? formatPace(prefs.threshold_pace) : "Not set";

  let prompt = buildTrainingRecommendationPrompt({
    athleteName: user.name,
    maxHr: prefs.max_hr ?? 190,
    restingHr: prefs.resting_hr ?? 50,
    thresholdPace: thresholdPaceText,
    ftp: ftpText,
    athleteProfile: user.profileSummary || "No profile summary available.",
    hrZones: hrZonesText,
    paceZones: paceZonesText,
    cyclingPowerZones: powerZonesText,
    recentActivities: activitiesText,
    weeklyDistance: weeklyDistance.toFixed(1),
    weeklyDuration: weeklyDuration.toFixed(1),
    weeklyAvgHr: weeklyAvgHr.toFixed(0),
    weeklyRuns,
    weeklyRides,
    weeklyOther,
    upcomingCompetitions: competitionsText,
    fixedPlan: fixedPlanText,
    weeklyTrimp: trimpLines.join("\n") || "No data.",
    recentRpe: rpeText,
    startDate,
    endDate,
    planningWeeks,
  });

  const allSports = ["running", "cycling", "swimming", "strength", "hiking", "rowing"];

  if (sportAvailability != null) {
    const selected = new Set(Object.keys(sportAvailability));
    const excluded = allSports.filter((s) => !selected.has(s));
    const lines: string[] = [];
    for (const [sport, cfg] of Object.entries(sportAvailability)) {
      const sportStart = cfg.start_date ?? startDate;
      if (sportStart <= startDate) {
        lines.push(`- ${sport}: available for the entire plan`);
      } else {
        lines.push(
          `- ${sport}: available from ${sportStart} onwards (do NOT schedule any ${sport} sessions before this date)`,
        );
      }
    }
    prompt += `\n\nIMPORTANT - Sport Availability Windows:\nThe athlete has specific availability dates for each sport. You MUST respect these windows:\n${lines.join("\n")}\nBefore a sport becomes available, use other available sports or rest days instead.`;
    if (excluded.length > 0) prompt += `\nDo NOT include sessions for: ${excluded.sort().join(", ")}.`;
  } else if (allowedSports != null) {
    const selected = new Set(allowedSports);
    if (selected.size > 0 && (selected.size !== allSports.length || ![...allSports].every((s) => selected.has(s)))) {
      if (selected.size === 1 && selected.has("running")) {
        prompt += `\n\nIMPORTANT: The athlete has requested a RUNNING-ONLY plan. Do NOT include cross-training sessions (cycling, swimming, strength, etc.). All sessions should have sport='running'. Use rest days instead of cross-training days.`;
      } else {
        const included = [...selected].sort().join(", ");
        const excluded = allSports.filter((s) => !selected.has(s)).sort().join(", ");
        prompt += `\n\nIMPORTANT: The athlete has selected specific sports for this plan. Only include sessions with sport set to one of: ${included}. Do NOT include sessions for: ${excluded}.`;
      }
    }
  }

  const result = await generateTrainingRecommendations(TRAINING_RECOMMENDATION_SYSTEM, prompt, onProgress);
  if (result.error) return { error: result.error };
  return result.data ?? {};
}

// ---------------------------------------------------------------------------
// convert_session
// ---------------------------------------------------------------------------

export async function convertSession(
  user: User,
  workout: Record<string, unknown>,
  targetType: string,
): Promise<Record<string, unknown>> {
  const prefs = (user.preferences ?? {}) as UserPreferences;
  const sourceType = targetType === "hr_based" ? "pace-based" : "HR-based";
  const prompt = buildPlanConversionPrompt({
    sourceType,
    targetType: targetType.replace(/_/g, " "),
    hrZones: formatHrZones(prefs.hr_zones),
    paceZones: formatPaceZones(prefs.pace_zones),
    sessionDetails: JSON.stringify(workout, null, 2),
    workoutType: typeof workout.type === "string" ? workout.type : "easy",
  });

  const result = await claudeConvert(PLAN_CONVERSION_SYSTEM, prompt);
  if (result.error) return { error: result.error };
  return result.data ?? {};
}

// ---------------------------------------------------------------------------
// save_recommendations
// ---------------------------------------------------------------------------

export async function saveRecommendations(
  user: User,
  recommendations: Record<string, unknown>,
): Promise<number> {
  const sessions = (recommendations.sessions ?? []) as unknown[];
  const today = new Date();
  today.setUTCHours(0, 0, 0, 0);
  const todayIso = isoOf(today);

  let saved = 0;
  for (const raw of sessions) {
    if (!raw || typeof raw !== "object" || Array.isArray(raw)) continue;
    const sessionData = raw as Record<string, unknown>;
    const dateValue = sessionData.date;
    if (typeof dateValue !== "string") continue;
    if (dateValue < todayIso) continue;

    const existing = await db
      .select({ id: trainingSessions.id })
      .from(trainingSessions)
      .where(
        and(
          eq(trainingSessions.userId, user.id),
          eq(trainingSessions.sessionDate, dateValue),
        ),
      )
      .limit(1);

    if (existing[0]) {
      await db
        .update(trainingSessions)
        .set({ recommendationWorkout: sessionData, updatedAt: new Date() })
        .where(eq(trainingSessions.id, existing[0].id));
    } else {
      await db.insert(trainingSessions).values({
        userId: user.id,
        sessionDate: dateValue,
        source: "app_recommendation",
        recommendationWorkout: sessionData,
      });
      saved += 1;
    }
  }
  return saved;
}
